relayer/src/link/relay_path.rs

```diff
@@ -748,48 +780,70 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
-            .is_ok()
-        {
-            return Ok(());
+    ) -> Result<Height, LinkError> {
+        if let Some(update_height) = Self::update_height(
+            self.src_chain(),
+            self.src_client_id().clone(),
+            dst_chain_height,
+        ) {
+            return Ok(update_height);
+        } else {
+            self.do_update_client_src(dst_chain_height, tracking_id, MAX_RETRIES)
         }
+    }
 
-        let mut src_err_ev = None;
-        for _ in 0..MAX_RETRIES {
-            let src_update = self.build_update_client_on_src(dst_chain_height)?;
-            info!(
-                "sending updateClient to client hosted on source chain for height {}",
-                dst_chain_height,
-            );
-
-            let tm = TrackedMsgs::new(src_update, tracking_id);
-
-            let src_tx_events = self
-                .src_chain()
-                .send_messages_and_wait_commit(tm)
-                .map_err(LinkError::relayer)?;
-            info!("result: {}", PrettyEvents(&src_tx_events));
-
-            src_err_ev = src_tx_events
-                .into_iter()
-                .find(|event| matches!(event, IbcEvent::ChainError(_)));
+    /// Perform actual update_client_src with retries.
+    ///
+    /// Note that the retry is only performed in the case when there
+    /// is a ChainError event. It would return error immediately if
+    /// there are other errors returned from calls such as
+    /// build_update_client_on_src.
+    fn do_update_client_src(
+        &self,
+        dst_chain_height: Height,
+        tracking_id: &str,
+        retries_left: usize,
+    ) -> Result<Height, LinkError> {
+        info!( "sending update_client to client hosted on source chain for height {} (retries left: {})", dst_chain_height, retries_left );
 
-            if src_err_ev.is_none() {
-                return Ok(());
+        let src_update = self.build_update_client_on_src(dst_chain_height)?;
+        let tm = TrackedMsgs::new(src_update, tracking_id);
+        let src_tx_events = self
+            .src_chain()
+            .send_messages_and_wait_commit(tm)
+            .map_err(LinkError::relayer)?;
+        info!("result: {}", PrettyEvents(&src_tx_events));
+
+        for event in src_tx_events {
+            match event {
+                err_event @ IbcEvent::ChainError(_) => {
+                    if retries_left == 0 {
+                        return Err(LinkError::client(ForeignClientError::chain_error_event(
+                            self.src_chain().id(),
+                            err_event,
+                        )));
+                    } else {
+                        return self.do_update_client_src(
+                            dst_chain_height,
+                            tracking_id,
+                            retries_left - 1,
+                        );
+                    }
+                }
+                IbcEvent::UpdateClient(ev) => return Ok(ev.height()),
```

Comment:
I think we need to make sure that this `UpdateClient` belongs to this particular client. (not sure if we have a check elsewhere for this)

Reply:
Seems like the tx can contain only `MsgUpdateClient`s for this client, so this shouldn't be a problem.